week8.tex

```diff
@@ -48,4 +53,220 @@ \subsection{Statics}
+  {S_1 \stepsTo<\alpha> S_1'}
+  {S_1 \parallel S_2 \stepsTo<\alpha> S_1' \parallel S_2}
+
+  \defrule[D-Par][dyn:par]
+  {S_1 \stepsTo<\alpha> S_1' \\
+    S_2 \stepsTo<\overline{\alpha}> S_2'}
+  {S_1 \parallel S_2 \stepsTo S_1' \parallel S_2'}
+
+\end{mathpar}
+Note that rule \ruleref{dyn:par} relies on the definition of $\overline{\alpha}$ representing the complementary action of $\alpha$. In this case, $\sndAc*{a}{s}$, the action to send $s$ over channel $a$, and $\rcvAc*{a}{s}$, the action to receive $s$ over channel $a$, are complementary to each other.
+
+The dynamics for I/O are relatively straightforward. If ${\cal I}$ is not empty and the program commands a \kw{read}, \ruleref{dyn:par} allows the overall configuration to take a step by pairing \ruleref{dyn:input} and \ruleref{dyn:read}. Similarly, if the program wishes to \kw{print}, \ruleref{dyn:output} and \ruleref{dyn:print} will apply concurrently.
+
+\ruleref{dyn:single} allows the program to take steps not pertaining to I/O according to the base CBPV dynamics. In this case $\alpha$ would be $\epsilon$, the silent action.
+
+\todo{Not sure how much of this to mention/is this extraneous detail? \ruleref{dyn:single} wasn't in the lecture, but I felt it was necessary to mention to show that the rest of the program can still run without needing to send messages}
+Note that this rule also allows for asynchronous sends since a process could perform an action $\sndAc*{a}{s}$ without a complementary receive. This is not an issue since we can view channels as a message queue, so any asynchronous send just adds its message to the queue. Reads must still be synchronous since there must be a value available on the queue to use in the rest of the program.
+
+\subsection{Equations}
+We present a few select equations:
+\begin{mathpar}
+
+  \defrule[E-Print-Fun][eq:print-fun]
+  {\Gamma \entails{\isOfTp{s}{\kw{string}}} \\
+    \Gamma, \isOfTp{x}{A} \entails{\isOfTp{C}{X}}}
+  {\Gamma \entails{\seqCmd*{\kw{print}(s)}{\lamEx*{x}{C} \equiv \lamEx*{x}{\seqCmd*{\kw{print}(s)}{C}}}} : \parrTy*{A}{X}}
+
+  \defrule[E-Read-Fun][eq:read-fun]
+  { \Gamma, \isOfTp{s}{\kw{string}}, \isOfTp{x}{A} \entails{\isOfTp{C}{X}}}
+  {\Gamma \entails{\fletEx*{\kw{read}}{s}{\lamEx*{x}{C} \equiv \lamEx*{x}{\fletEx*{\kw{read}}{s}{C}}}} : \parrTy*{A}{X}}
+
+  \defrule[E-Print-Bind][eq:print-bind]
+  {\Gamma \entails{\isOfTp{s}{\kw{string}}} \\
+   \Gamma \entails{\isOfTp{C_1}{\freeTy{A_1}}} \\
+   \Gamma, \isOfTp{x}{A_1} \entails{\isOfTp{C_2}{X_2}}}
+  {\Gamma \entails{\fletEx*{\left( \seqCmd*{\kw{print}(s)}{C_1} \right)}{x}{C_2} \equiv \seqCmd*{\kw{print}(s)}{\fletEx*{C_1}{x}{C_2}}} : X_2}
+
+  \defrule[E-Read-Bind][eq:print-bind]
+  {\Gamma, \isOfTp{s}{\kw{string}} \entails{\isOfTp{C_1}{\freeTy{A_1}}} \\
+   \Gamma, \isOfTp{x}{A_1} \entails{\isOfTp{C_2}{X_2}}}
+  {\Gamma \entails{\fletEx*{\left( \fletEx*{\kw{read}}{s}{C_1} \right)}{x}{C_2} \equiv \fletEx*{\kw{read}}{s}{\fletEx*{C_1}{x}{C_2}}} : X_2}
+
+\end{mathpar}
```

Comment:
This is a lot of rules. Care to put them in a figure? You can look at week8 branch.